pbft/pbft.cpp

```diff
@@ -0,0 +1,363 @@
+
+    //TODO: keep track of what requests we've seen based on timestamp and only send preprepares once - KEP-329
+
+    uint64_t request_seq = this->next_issued_sequence_number++;
+    pbft_operation& op = this->find_operation(this->view, request_seq, msg.request());
+
+    this->do_preprepare(op);
+}
+
+void
+pbft::handle_preprepare(const pbft_msg& msg)
+{
+
+    // If we've already accepted a preprepare for this view+sequence, and it's not this one, then we should reject this one
+    // Note that if we get the same preprepare more than once, we can still accept it
+    const log_key_t log_key(msg.view(), msg.sequence());
+
+    if (auto lookup = this->accepted_preprepares.find(log_key);
+        lookup != this->accepted_preprepares.end()
+        && std::get<2>(lookup->second) != pbft_operation::request_hash(msg.request()))
+    {
+
+        LOG(debug) << "Rejecting preprepare because I've already accepted a conflicting one \n";
+        return;
+    }
+    else
+    {
+        pbft_operation& op = this->find_operation(msg);
+        op.record_preprepare();
+
+        // This assignment will be redundant if we've seen this preprepare before, but that's fine
+        accepted_preprepares[log_key] = op.get_operation_key();
+
+        this->do_preprepared(op);
+        this->maybe_advance_operation_state(op);
+    }
+}
+
+void
+pbft::handle_prepare(const pbft_msg& msg)
+{
+
+    // Prepare messages are never rejected, assuming the sanity checks passed
+    pbft_operation& op = this->find_operation(msg);
+
+    op.record_prepare(msg);
+    this->maybe_advance_operation_state(op);
+}
+
+void
+pbft::handle_commit(const pbft_msg& msg)
+{
+
+    // Commit messages are never rejected, assuming  the sanity checks passed
+    pbft_operation& op = this->find_operation(msg);
+
+    op.record_commit(msg);
+    this->maybe_advance_operation_state(op);
+}
+
+void
+pbft::broadcast(const pbft_msg& msg)
+{
+    auto json_ptr = std::make_shared<bzn::message>(this->wrap_message(msg));
+
+    for (const auto& peer : this->peers)
+    {
+        this->node->send_message(make_endpoint(peer), json_ptr);
+    }
+}
+
+void
+pbft::maybe_advance_operation_state(pbft_operation& op)
+{
+    if (op.get_state() == pbft_operation_state::prepare && op.is_prepared())
+    {
+        this->do_prepared(op);
+    }
+
+    if (op.get_state() == pbft_operation_state::commit && op.is_committed())
+    {
+        this->do_committed(op);
+    }
+}
+
+pbft_msg
+pbft::common_message_setup(const pbft_operation& op)
+{
+    pbft_msg msg;
+    msg.set_view(op.view);
+    msg.set_sequence(op.sequence);
+    msg.set_allocated_request(new pbft_request(op.request));
+
+    // Some message types don't need this, but it's cleaner to always include it
+    msg.set_sender(this->uuid);
+
+    return msg;
+}
+
+void
+pbft::do_preprepare(pbft_operation& op)
+{
+    LOG(debug) << "Doing preprepare for operation " << op.debug_string();
+
+    pbft_msg msg = this->common_message_setup(op);
```

Comment:
perhaps send in message type to helper function so you can remove all of the msg.set_type calls?